### ISSUE TITLE
If target_folder is specified, epinow fails unless reported_cases is a data.table

### ISSUE THREAD

**Issue by ellisp** [epiforecasts/EpiNow2#35]
A call to `epinow` with the data in a data frame that works fine if you aren't saving results will cause an error if you try to save the results by specifying `target_folder`.  Reproducible example:

```R
library(EpiNow2)
## Construct example distributions
generation_time <- list(mean = EpiNow2::covid_generation_times[1, ]$mean,
                        mean_sd = EpiNow2::covid_generation_times[1, ]$mean_sd,
                        sd = EpiNow2::covid_generation_times[1, ]$sd,
                        sd_sd = EpiNow2::covid_generation_times[1, ]$sd_sd,
                        max = 30)

incubation_period <- list(mean = EpiNow2::covid_incubation_period[1, ]$mean,
                          mean_sd = EpiNow2::covid_incubation_period[1, ]$mean_sd,
                          sd = EpiNow2::covid_incubation_period[1, ]$sd,
                          sd_sd = EpiNow2::covid_incubation_period[1, ]$sd_sd,
                          max = 30)

reporting_delay <- EpiNow2::bootstrapped_dist_fit(rlnorm(100, log(6), 1))
## Set max allowed delay to 60 days to truncate computation
reporting_delay$max <- 60

## Example case data
reported_cases <- as.data.frame(EpiNow2::example_confirmed[1:40] )

## This works:
out <- epinow(reported_cases = reported_cases, generation_time = generation_time,
              delays = list(incubation_period, reporting_delay),
              rt_prior = list(mean = 1, sd = 1),
              samples = 1000, warmup = 200, cores = 1, chains = 1,
              verbose = TRUE, return_fit = TRUE)

## This doesn't:
out <- epinow(reported_cases = reported_cases, generation_time = generation_time,
              delays = list(incubation_period, reporting_delay),
              rt_prior = list(mean = 1, sd = 1),
              samples = 1000, warmup = 200, cores = 1, chains = 1,
              verbose = TRUE, return_fit = TRUE, target_folder = 'saved-results')

## This works too:
out <- epinow(reported_cases = as.data.table(reported_cases), generation_time = generation_time,
              delays = list(incubation_period, reporting_delay),
              rt_prior = list(mean = 1, sd = 1),
              samples = 1000, warmup = 200, cores = 1, chains = 1,
              verbose = TRUE, return_fit = TRUE, target_folder = 'saved-results')

```

This is the error:

```
Error in `[.data.frame`(reported_cases, confirm > 0) : 
  object 'confirm' not found

> traceback()
7: `[.data.frame`(reported_cases, confirm > 0)
6: reported_cases[confirm > 0]
5: epinow(reported_cases = reported_cases, generation_time = generation_time, 
       delays = list(incubation_period, reporting_delay), rt_prior = list(mean = 1, 
           sd = 1), samples = 1000, warmup = 200, cores = 1, chains = 1, 
       verbose = TRUE, return_fit = TRUE, target_folder = "saved-results") at .active-rstudio-document#23
4: eval(ei, envir)
3: eval(ei, envir)
2: withVisible(eval(ei, envir))
1: source("~/.active-rstudio-document", echo = TRUE)
```

This could probably be fixed with an appropriate `as.data.table()` somewhere but I will leave it to you whether you want to do that, or just specify that reported_cases must be a data.table (I would vote you allow data frames and tibbles but that's your call).

**Comment by ellisp** [comment]
I think this expectation of the data being a data.table might also cause problems with the plot method too, irrespective of whether there is a target_folder. I have found as of a version of `EpiNow2` I installed today that Stan finishes its model fitting ok and then I get an error message:

```R
Error in `>=.default`(date, min(estimate$date, na.rm = TRUE)) : 
  comparison (5) is possible only for atomic and list types
```

I have just noticed that this problem seems to go away if my original data is a data.table.

**Comment by seabbs** [comment]
Thanks for this @ellisp - good spot. 

An internal conversion to `data.table` is a good idea (and to be honest something I thought I had implemented!). 

Very odd that this is only an issue when saving results.

**Comment by seabbs** [comment]
Should be fixed in #37
